components/esp_jpeg/jpeg_decoder.c

```diff
@@ -0,0 +1,219 @@
+#include "rom/tjpgd.h"
+#else
+#error Using JPEG decoder from ROM is not supported for selected target. Please select external code in menuconfig.
+#endif
+#else
+/* When Tiny JPG Decoder is not in ROM or selected external code */
+#include "tjpgd.h"
+#endif
+
+static const char *TAG = "JPEG";
+
+#define LOBYTE(u16)     ((uint8_t)(((uint16_t)(u16)) & 0xff))
+#define HIBYTE(u16)     ((uint8_t)((((uint16_t)(u16))>>8) & 0xff))
+
+#if defined(JD_FASTDECODE) && (JD_FASTDECODE == 2)
+#define JPEG_WORK_BUF_SIZE  65472
+#else
+#define JPEG_WORK_BUF_SIZE  5000
+#endif
+
+/* If not set JD_FORMAT, it is set in ROM to RGB888, otherwise, it can be set in config */
+#ifndef JD_FORMAT
+#define JD_FORMAT 0
+#endif
+
+/* Output color bytes from tjpgd (depends on JD_FORMAT) */
+#if (JD_FORMAT==0)
+#define ESP_JPEG_COLOR_BYTES    3
+#elif  (JD_FORMAT==1)
+#define ESP_JPEG_COLOR_BYTES    2
+#elif  (JD_FORMAT==2)
+#error Grayscale image output format is not supported
+#define ESP_JPEG_COLOR_BYTES    1
+#endif
+
+/*******************************************************************************
+* Function definitions
+*******************************************************************************/
+static uint8_t jpeg_get_div_by_scale(esp_jpeg_image_scale_t scale);
+static uint8_t jpeg_get_color_bytes(esp_jpeg_image_format_t format);
+
+static unsigned int jpeg_decode_in_cb(JDEC *jd, uint8_t *buff, unsigned int nbyte);
+static unsigned int jpeg_decode_out_cb(JDEC *jd, void *bitmap, JRECT *rect);
+/*******************************************************************************
+* Public API functions
+*******************************************************************************/
+
+esp_err_t esp_jpeg_decode(esp_jpeg_image_cfg_t *cfg, esp_jpeg_image_output_t *img)
+{
+    esp_err_t ret = ESP_OK;
+    uint8_t *workbuf = NULL;
+    JRESULT res;
+    JDEC JDEC;
+
+    assert(cfg != NULL);
+    assert(img != NULL);
+
+    workbuf = heap_caps_malloc(JPEG_WORK_BUF_SIZE, MALLOC_CAP_DEFAULT);
+    ESP_GOTO_ON_FALSE(workbuf, ESP_ERR_NO_MEM, err, TAG, "no mem for JPEG work buffer");
+
+    cfg->priv.read = 0;
+
+    /* Prepare image */
+    res = jd_prepare(&JDEC, jpeg_decode_in_cb, workbuf, JPEG_WORK_BUF_SIZE, cfg);
+    ESP_GOTO_ON_FALSE((res == JDR_OK), ESP_FAIL, err, TAG, "Error in preparing JPEG image!");
+
+    uint8_t scale_div = jpeg_get_div_by_scale(cfg->out_scale);
+    uint8_t out_color_bytes = jpeg_get_color_bytes(cfg->out_format);
+
+    /* Size of output image */
+    uint32_t outsize = (JDEC.height / scale_div) * (JDEC.width / scale_div) * out_color_bytes;
+    ESP_GOTO_ON_FALSE((outsize <= cfg->outbuf_size), ESP_ERR_NO_MEM, err, TAG, "Not enough size in output buffer!");
+
+    /* Size of output image */
+    img->height = JDEC.height / scale_div;
+    img->width = JDEC.width / scale_div;
+
+    /* Decode JPEG */
+    res = jd_decomp(&JDEC, jpeg_decode_out_cb, cfg->out_scale);
```

Comment:
I'm getting some compilation warnings:
```
C:/github_repos/esp-bsp/components/esp_jpeg/jpeg_decoder.c:97:28: warning: passing argument 2 of 'jd_decomp' from incompatible pointer type [-Wincompatible-pointer-types]
     res = jd_decomp(&JDEC, jpeg_decode_out_cb, cfg->out_scale);
                            ^~~~~~~~~~~~~~~~~~
In file included from C:/github_repos/esp-bsp/components/esp_jpeg/jpeg_decoder.c:25:
C:/github_repos/esp-bsp/components/esp_jpeg/tjpgd/tjpgd.h:95:36: note: expected 'int (*)(JDEC *, void *, JRECT *)' {aka 'int (*)(struct JDEC *, void *, struct <anonymous> *)'} but argument is of type 'unsigned int (*)(JDEC *, void *, JRECT *)' {aka 'unsigned int (*)(struct JDEC *, void *, struct <anonymous> *)'}
 JRESULT jd_decomp (JDEC *jd, int (*outfunc)(JDEC *, void *, JRECT *), uint8_t scale);
                              ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```

Reply:
The TJPG Decoder in ROM is little different than TJPG decoder code in component. There is newer version in component. I didn't want to add more #if-def for this warning.